// @gear-js/api/lib/wasm/WasmMeta.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWasmMetadata = void 0;
const importObj_1 = __importDefault(require("./importObj"));
const utils_1 = require("./utils");
function getWasmMetadata(wasmBytes, showDebug = false) {
    return __awaiter(this, void 0, void 0, function* () {
        const memory = new WebAssembly.Memory({ initial: 256 });
        const module = yield WebAssembly.instantiate(wasmBytes, (0, importObj_1.default)(memory, showDebug));
        const { exports } = module.instance;
        return exports
            ? {
                types: `0x${(0, utils_1.readMetaValue)(memory, exports.meta_registry)}`,
                init_input: (0, utils_1.readMetaValue)(memory, exports.meta_init_input),
                init_output: (0, utils_1.readMetaValue)(memory, exports.meta_init_output),
                async_init_input: (0, utils_1.readMetaValue)(memory, exports.meta_async_init_input),
                async_init_output: (0, utils_1.readMetaValue)(memory, exports.meta_async_init_output),
                handle_input: (0, utils_1.readMetaValue)(memory, exports.meta_handle_input),
                handle_output: (0, utils_1.readMetaValue)(memory, exports.meta_handle_output),
                async_handle_input: (0, utils_1.readMetaValue)(memory, exports.meta_async_handle_input),
                async_handle_output: (0, utils_1.readMetaValue)(memory, exports.meta_async_handle_output),
                title: (0, utils_1.readMetaValue)(memory, exports.meta_title),
                meta_state_input: (0, utils_1.readMetaValue)(memory, exports.meta_state_input),
                meta_state_output: (0, utils_1.readMetaValue)(memory, exports.meta_state_output),
            }
            : {};
    });
}
exports.getWasmMetadata = getWasmMetadata;
