// @gear-js/api/lib/create-type/generate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("../utils");
const getTree = (type, name, value, count) => {
    const tree = { type, name, value: (0, utils_1.isJSON)(value) ? (0, utils_1.toJSON)(value) : value };
    return type === 'Array' ? Object.assign(Object.assign({}, tree), { count }) : tree;
};
exports.default = {
    Primitive: (type) => {
        return getTree('Primitive', type, type);
    },
    Tuple: (name, type) => {
        return getTree('Tuple', name, type);
    },
    Struct: (name, value) => {
        return getTree('Struct', name, value);
    },
    Enum: (name, options) => {
        return getTree('Enum', name, options);
    },
    Array: (name, type, count) => {
        return getTree('Array', name, type, count);
    },
    Option: (name, some) => {
        return getTree('Option', name, some);
    },
    Result: (name, ok, err) => {
        return getTree('Result', name, { ok, err });
    },
    Vec: (name, type) => {
        return getTree('Vec', name, type);
    },
    BTreeMap: (name, key, value) => {
        return getTree('BTreeMap', name, { key, value });
    },
    BTreeSet: (name, type) => {
        return getTree('BTreeSet', name, type);
    },
};
